zebra-chain/src/orchard/tree.rs

```diff
@@ -42,9 +42,13 @@ fn merkle_crh_orchard(layer: u8, left: [u8; 32], right: [u8; 32]) -> [u8; 32] {
     let mut s = bitvec![Lsb0, u8;];
 
     // Prefix: l = I2LEBSP_10(MerkleDepth^Orchard − 1 − layer)
-    s.extend_from_slice(&layer.bits::<Lsb0>()[0..10]);
-    s.extend_from_slice(&left.bits::<Lsb0>()[0..255]);
-    s.extend_from_slice(&right.bits::<Lsb0>()[0..255]);
+    s.extend_from_bitslice(&BitSlice::<Lsb0, _>::from_element(&layer)[0..10]);
```

Comment:
This looks weird. It seems to me that this creates an 8-bit slice and tries to read 10 bits from it :thinking: Not sure if this is meant to pad the `layer` input with two zero bits:

```suggestion
    s.extend_from_bitslice(&BitArray::<Lsb0, _>::from([layer, 0])[0..10]);
```

or if `layer` should be another type, like `&BitSlice` or `u16` (and here's hoping one day we have something like `BitArray<10>` :crossed_fingers:)

Reply:
I think this is a question for @dconnolly ^

I remember we discussed it during the original PR review.

Reply:
Hmm yeah the equivalent non-slice code panics with an out-of-range error:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=970fcf9d20c1b91c1307fdecd253e939

```rust
fn main() {
    let arry = [0, 1];
    let out_of_range = &arry[1..5];
    println!("{:?}", out_of_range);
}
```

Reply:
Un-resolving this until @dconnolly has checked the code.

Reply:
> This looks weird. It seems to me that this creates an 8-bit slice and tries to read 10 bits from it 🤔 Not sure if this is meant to pad the `layer` input with two zero bits:
> 
> or if `layer` should be another type, like `&BitSlice` or `u16` (and here's hoping one day we have something like `BitArray<10>` 🤞)

Yes this is basically padded out to 10 bits always. This max value of 32 (the depth/height of the note commitment tree) is padded out to 10 bits to function as a prefix for the hash function, providing domain separation between inputs at different layers of the tree. https://zips.z.cash/protocol/protocol.pdf#orchardmerklecrh 
The prefix `l` is needed to be a `I2LEBSP` encoding, where `I2LEBSP_l(x)`  is the sequence of ℓ bits representing x in little-endian order. https://zips.z.cash/protocol/protocol.pdf#endian

Reply:
I think you are right @jvff